### PR TITLE
PFCP: Wrong Bridge ID format in TSN Bridge ID attribute

### PR DESCRIPTION
**Pull request by eric5800602**
Modify tsn-bridge-id and related test files to follow 29.244 v16
The TSN Bridge ID is defined in IEEE 802.1Q [30] clause 14.2.5 and **the value shall be encoded as an Unsigned64 binary integer.**
Please check the following Wireshark issue:
https://gitlab.com/wireshark/wireshark/-/issues/18760

### CODE REVIEW

File: message/message_test.go [hunk 1]
```diff
@@ -7,10 +7,11 @@ package message_test
...
 var (
-	mac1, _ = net.ParseMAC("12:34:56:78:90:01")
-	mac2, _ = net.ParseMAC("12:34:56:78:90:02")
-	mac3, _ = net.ParseMAC("12:34:56:78:90:03")
-	mac4, _ = net.ParseMAC("12:34:56:78:90:04")
+	mac1, _           = net.ParseMAC("12:34:56:78:90:01")
+	mac2, _           = net.ParseMAC("12:34:56:78:90:02")
+	mac3, _           = net.ParseMAC("12:34:56:78:90:03")
+	mac4, _           = net.ParseMAC("12:34:56:78:90:04")
+	tsn_bridge_mac, _ = net.ParseMAC("12:34:56:78:90:05:67:89")
```
Comment (wmnsk) [inline message/message_test.go:14]: Please don't use snake case for a variable.

File: ie/tsn-bridge-id.go [hunk 2]
```diff
@@ -15,9 +15,9 @@ func NewTSNBridgeID(mac net.HardwareAddr) *IE {
 		return New(TSNBridgeID, []byte{0x00})
 	}
 
-	b := make([]byte, 7)
+	b := make([]byte, 9)
 	b[0] = 0x01
-	copy(b[1:7], mac)
+	copy(b[1:9], mac)
 	return New(TSNBridgeID, b)
```
Comment (wmnsk) [inline ie/tsn-bridge-id.go:21]: By using the length of the input, we can still support constructing with the 6-byte MAC address. Not quite sure if it's worth, but it's basically a good idea to keep the old implementation as long as possible.

```go
l := len(mac) + 1
b := make([]byte, l)
...
```

File: ie/tsn-bridge-id.go [hunk 3]
```diff
@@ -40,10 +40,10 @@ func (i *IE) TSNBridgeID() (net.HardwareAddr, error) {
...
 	case TSNBridgeID:
 		if has1stBit(i.Payload[0]) {
-			if len(i.Payload) < 7 {
+			if len(i.Payload) < 9 {
 				return nil, io.ErrUnexpectedEOF
 			}
-			return net.HardwareAddr(i.Payload[1:7]), nil
+			return net.HardwareAddr(i.Payload[1:9]), nil
 		}
 		return nil, nil
```
Comment (wmnsk) [inline ie/tsn-bridge-id.go:48]: Can we do something like this?

```go
		if !has1stBit(i.Payload[0]) {
			return nil, nil
		}

		if len(i.Payload) < 9 {
			if len(i.Payload) < 7 {
				return nil, io.ErrUnexpectedEOF
			}
			return net.HardwareAddr(i.Payload[1:7]), nil 
		}
		return net.HardwareAddr(i.Payload[1:9]), nil
```

**Review by wmnsk — COMMENTED** [review]
Thanks!
If I understand correctly, the spec has changed at some time (long ago though), so I'd like to support both if possible - please check inline comments. And it seems some tests still need to be updated.

File: message/message_test.go (hunk 1 above)
Comment (eric5800602) [reply message/message_test.go]: Sorry, I change the variable name to upper camel case.

File: message/message_test.go [hunk 4]
```diff
@@ -7,10 +7,11 @@ package message_test
...
 var (
-	mac1, _ = net.ParseMAC("12:34:56:78:90:01")
-	mac2, _ = net.ParseMAC("12:34:56:78:90:02")
-	mac3, _ = net.ParseMAC("12:34:56:78:90:03")
-	mac4, _ = net.ParseMAC("12:34:56:78:90:04")
+	mac1, _         = net.ParseMAC("12:34:56:78:90:01")
+	mac2, _         = net.ParseMAC("12:34:56:78:90:02")
+	mac3, _         = net.ParseMAC("12:34:56:78:90:03")
+	mac4, _         = net.ParseMAC("12:34:56:78:90:04")
+	TsnBridgeMac, _ = net.ParseMAC("12:34:56:78:90:05:67:89")
```
Comment (wmnsk) [inline message/message_test.go:14]: This should not be exported. https://go.dev/doc/effective_go#names

```suggestion
	tsnBridgeMac, _ = net.ParseMAC("12:34:56:78:90:05:67:89")
```

**Review by wmnsk — COMMENTED** [review]
Please check the naming again - a variable name should start with lower case unless it's used outside the package.

File: ie/tsn-bridge-id.go [hunk 5]
```diff
@@ -39,13 +39,17 @@ func (i *IE) TSNBridgeID() (net.HardwareAddr, error) {
...
+		}
+
+		if len(i.Payload) < 9 {
 			if len(i.Payload) < 7 {
 				return nil, io.ErrUnexpectedEOF
 			}
 			return net.HardwareAddr(i.Payload[1:7]), nil
 		}
-		return nil, nil
+		return net.HardwareAddr(i.Payload[1:9]), nil
```
Comment (louisroyer) [inline ie/tsn-bridge-id.go:52]: I would suggest the following instead, to be able to also raise an error on malformed 8-bytes length IEs:
```golang
case TSNBridgeID:
	if !has1stBit(i.Payload[0]) {
		return nil, nil
	}
	// If the first bit is set to 1,
	// the payload contains
	// 1. a first byte, and either
	// 2a. exactly 6 bytes, or
	// 2b. at least 8 bytes.
	if (len(i.Payload) != 7) && (len(i.Payload) < 9) {
		return nil, ie.ErrUnexpectedEOF
	}
	// In 129.244 V16.2.0 to V16.3.1,
	// MAC address shall be encoded
	// as 12-digit hexadecimal numbers
	if len(i.Payload) == 7 {
		return net.HardwareAddr(i.Payload[1:7]), nil)
	}
	// Starting from V16.4.0, 
	// the User Plane Node Value, shall be encoded
	// as an IEEE 801Q Uint64
	return net.HardwareAddr(i.Payload[1:9]), nil)
```

**Review by louisroyer — CHANGES_REQUESTED** [review]

File: ie/tsn-bridge-id.go (hunk 5 above)
Comment (wmnsk) [reply ie/tsn-bridge-id.go:52]: that's neat, would take it! @eric5800602 could you update?

Comment (eric5800602) [reply ie/tsn-bridge-id.go:52]: @wmnsk Sure.
Thanks for the suggestion.

**Pull request merged**